rbac/internal/utils.py

```
@@ -900,25 +900,84 @@ def clean_invalid_workspace_resource_definitions(dry_run: bool = False) -> dict:
+            # Race: PK taken, or another ungrouped-hosts row for this tenant was created.
+            raced = (
+                Workspace.objects.select_for_update()
+                .filter(tenant=tenant, type=Workspace.Types.UNGROUPED_HOSTS)
+                .first()
+            )
+            if raced is not None:
+                if raced.id != workspace_id:
+                    raise UngroupedWorkspaceError(
+                        f"Tenant org_id={tenant.org_id} already has ungrouped-hosts workspace "
+                        f"{raced.id}, which does not match requested id {workspace_id}."
+                    )
```

Ruff (0.15.21)

[warning] Within an `except` clause, raise exceptions with `raise ... from err` or `raise ... from None` to distinguish them from errors in exception handling

(B904)